Rover/JetsonGStreamerServer/gserver.cpp

```diff
@@ -0,0 +1,227 @@
+  for (int i = 0; i < n_fields; i++) 
+  {
+    const gchar *name;
+    name = gst_structure_nth_field_name(device, i);
+    g_print("%d field name: %s\n", i, name);
+  }
+}
+
+gboolean bus_func (GstBus * bus, GstMessage * message, gpointer user_data)
+{
+   GstDevice *devi;
+   gchar *name;
+
+   switch (GST_MESSAGE_TYPE (message)) 
+   {
+     case GST_MESSAGE_DEVICE_ADDED:
+       GstStructure *type;
+
+       gst_message_parse_device_added (message, &devi);
+       name = gst_device_get_display_name (devi);
+       type = gst_device_get_properties(devi);
+       g_print("Device added: %s\n", name);
+       g_free (name);
+       g_print("Device path: %s\n", gst_structure_get_string(type, "device.path"));
+       structure_fields(type);
+       gst_structure_free(type);
+       gst_object_unref (devi);
+       break;
+     case GST_MESSAGE_DEVICE_REMOVED:
+       gst_message_parse_device_removed (message, &devi);
+       name = gst_device_get_display_name (devi);
+       g_print("Device removed: %s\n", name);
+       g_free (name);
+       gst_object_unref (devi);
+       break;
+     default:
+       break;
+   }
+
+   return G_SOURCE_CONTINUE;
+}
+
+// v4l2src device="/dev/video0" ! 
+// "video/x-raw, format=(string)I420, width=(int)1280, height=(int)720" ! 
+// omxh264enc ! 
+// 'video/x-h264, stream-format=(string)byte-stream' ! 
+// h264parse ! 
+// rtph264pay ! 
+// udpsink host=192.168.0.5 port=5555
+GstElement * start_device(GstDevice *dev, const char *client) 
+{
+  GstStructure *str;
+  gchar *name, *cl;
+  GstElement *pipeline, *source, *depay, *encoder, *parse, *sink;
+  GstElement *s_cap, *e_cap, *scale, *scale_cap;
+  GstCaps *s_cap_unapplied, *e_cap_unapplied, *scale_cap_unapplied;
+
+  name = gst_device_get_display_name(dev);
+  cl = gst_device_get_device_class(dev);
+  str = gst_device_get_properties(dev);
+  printf("device name: %s\n", name);
+  printf("device class: %s\n", cl);
+  g_print("device path: %s\n", gst_structure_get_string(str, "device.path")); 
+  
+  // setting elements of pipeline
+  pipeline = gst_pipeline_new(gst_device_get_display_name(dev));
+  source = gst_element_factory_make("v4l2src", NULL);
+  s_cap = gst_element_factory_make("capsfilter", NULL);
+  e_cap = gst_element_factory_make("capsfilter", NULL);
+  depay = gst_element_factory_make("rtph264pay", NULL);
+  encoder = gst_element_factory_make("omxh264enc", NULL);
+  parse = gst_element_factory_make("h264parse", NULL);
+  sink = gst_element_factory_make("udpsink", NULL); 
+  scale = gst_element_factory_make("videoscale", NULL);
+  scale_cap = gst_element_factory_make("capsfilter", NULL);
+
+  // options for the source
+  g_object_set(G_OBJECT(source), "device", gst_structure_get_string(str, "device.path")
+                                                                                , NULL);
+  std::unordered_map<std::string, std::tuple<gint, gint, gint, gint, gint> > table;
+  table["RICOH THETA S"] = std::make_tuple(1280, 720, 5555, 640, 360);
```

Comment:
Maybe add a comment or use variable names here to explain what these parameter names are